cpp/src/gandiva/simple_arena.h

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef GANDIVA_SIMPLE_ARENA_H
+#define GANDIVA_SIMPLE_ARENA_H
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+namespace gandiva {
+
+#ifdef GDV_HELPERS
+namespace helpers {
+#endif
+
+/// \brief Simple arena allocator.
+///
+/// Memory is allocated from system in units of chunk-size, and dished out in the
+/// requested sizes. If the requested size > chunk-size, allocate directly from the
+/// system.
+///
+/// The allocated memory gets released only when the arena is destroyed, or on
+/// Reset.
+///
+/// This code is not multi-thread safe, and avoids all locking for efficiency.
+///
+class SimpleArena {
+ public:
+  explicit SimpleArena(int32_t chunk_size = 4096);
+
+  // Allocate buffer of requested size.
+  uint8_t* Allocate(int32_t size);
+
+  // Reset arena state.
+  void Reset();
+
+  // total bytes allocated from system.
+  int32_t total_bytes() { return total_bytes_; }
+
+  // total bytes available for allocations.
+  int32_t avail_bytes() { return avail_bytes_; }
+
+ private:
+  // Allocate new chunk.
+  void AllocateChunk(int size);
+
+  // The chunk-size used for allocations from system.
+  int32_t chunk_size_;
+
+  // The size of the first chunk.
+  int32_t first_chunk_size_;
+
+  // Total bytes allocated from system.
+  int32_t total_bytes_;
+
+  // Bytes available from allocated chunk.
+  int32_t avail_bytes_;
+
+  // buffer from current chunk.
+  uint8_t* avail_buf_;
+
+  // List of alloced chunks.
+  std::vector<std::unique_ptr<uint8_t>> chunks_;
+};
+
+inline SimpleArena::SimpleArena(int32_t chunk_size)
+    : chunk_size_(chunk_size),
+      first_chunk_size_(0),
+      total_bytes_(0),
+      avail_bytes_(0),
+      avail_buf_(NULL) {}
+
+inline uint8_t* SimpleArena::Allocate(int32_t size) {
+  if (avail_bytes_ < size) {
+    AllocateChunk(size <= chunk_size_ ? chunk_size_ : size);
+  }
+
+  uint8_t* ret = avail_buf_;
+  avail_buf_ += size;
+  avail_bytes_ -= size;
+  return ret;
+}
+
+inline void SimpleArena::AllocateChunk(int size) {
+  std::unique_ptr<uint8_t> chunk(new uint8_t[size]);
```

Comment:
Arrow has the MemoryPool abstraction, which will often use jemalloc under the hood. We probably shouldn't hardcode calls to the C++ allocator for performance-sensitive allocations.

Reply:
Probably would need to change packaging(include the jemalloc library in static version of gandiva) too if we use MemoryPool?

Reply:
I think jemalloc is already linked statically into Arrow, so you shouldn't have anything special to do here?

Reply:
ah ok, we found that some of the dependencies were not linked/exported in the static version of arrow (for e.g. boost) when we were doing the packaging work for Gandiva.

Thanks! that would be good if it is already linked statically.

Reply:
There are build options for some of these things, like Boost, since not everyone wants to statically link. With jemalloc it is a private version so always statically linked.

Reply:
@pitrou @wesm - this arena is being used for very transient allocations. The allocations are all released at the end of processing a batch. Is it okay to use default_memory_pool() here (instead of adding a pool parameter to all our APIs) ?

Reply:
The pool could be part of the execution context like the arena, no?

Reply:
> The pool could be part of the execution context like the arena, no?

yeah, but I need to get this all the way from the Project::Evaluate() and Filter::Evaluate. Some of these APIs don't take a pool parameter now (since they don't do any allocations). So, their signatures need to change. For now, I've used the default_memory_pool. I'll create a new jira to fix all the Evaluate API signatures to take the pool as a parameter. Is that ok ?

Reply:
Yes, that's fine with me. You may also want to check if that changes performance.